README.md
# Placevent3

Placevent3 - Python 3 translation of Placevent modified
by Shinji Aono (aono@rist.or.jp) and Michio Katouda (katouda@rist.or.jp)    

Placevent - 3D-RISM-based solvent and ion placement software created 
by Daniel Sindhikara (sindhikara@gmail.com)    
This program is designed to automatically place explicit solvent atoms/ions based 
on 3D-RISM data. Placevent is compatible with AMBER .dx files and MDF .h5 files.    
https://github.com/dansind/Placevent

The details of the algorithm are described in:    
Placevent: An algorithm for prediction of explicit solvent atom distribution 
-- Awpplication to HIV-1 protease and F-ATP synthase    
Daniel J. Sindhikara, Norio Yoshida, Fumio Hirata    
http://dansindhikara.com/Software/Entries/2012/6/22_Placevent_New.html    

If you have Numpy, this package should work out of the box. Please contact authors if you have problems.    
This package requires Numpy.    

## EXAMPLE:
Please see the tests/1L2Y/ for examples/tests    
In that directory, run test.sh to see if your local installation matches the reference output.    

## TUTORIAL:
See my website: http://dansindhikara.com/Tutorials/Entries/2012/1/1_Using_3D-RISM_and_PLACEVENT.html
for a tutorial.    


g(r)_0 is printed in the occupation column    
g(r)_i is printed in the beta column    



    Copyright (C) 2012 Daniel J. Sindhikara    

    This program is free software: you can redistribute it and/or modify    
    it under the terms of the GNU General Public License as published by    
    the Free Software Foundation, either version 3 of the License, or    
    (at your option) any later version.    

    This program is distributed in the hope that it will be useful,    
    but WITHOUT ANY WARRANTY; without even the implied warranty of    
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    
    GNU General Public License for more details.    

    You should have received a copy of the GNU General Public License    
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGES.md
# Placevent3

Oct 9, 2023
Upgrade to 0.1
1) Add grid.py to package
2) Port from python 2 to python 3.

# Placevent
Oct 9, 2013
Upgrade to 1.4
1) Added Molecular Design Frontier 3D-RISM (MDF 3D-RISM) .h5 compatibility.
2) Use grid.py instead of internal pgrid module

Mar 23, 2013
Upgrade to 1.3
1) Steal population from pseudobulk (outside grid) instead of failing. 
Allows placement of many more atoms within smaller grid.

Mar 20, 2013
Upgrade to 1.2.2
1) Modified grid.py to read slightly differently formatted .dx files

Jan 28, 2013 
Upgrade to 1.2

1) Typo: Fixed minor error in equation for evacuation of final shell (will give slightly different results)
2) Adapted to setup.py-based installation 
    a) Added setup.py
    b) Rewrote tests to allow for either installed or unzipped placevent
    c) Uploaded to Pypi (try pip install placevent) 
       


tests/2ZYE/run_predict_watOs.sh
#!/bin/bash

#pwd
guv=complex_r3d-kh_guv.O.0.dx
thresh_p=5.0
watOs_pdb=watOs_placevent5.0.pdb

placevent ${guv} 55.5 ${thresh_p} 2.4 > ${watOs_pdb}


Placevent3/pdb.py
'''
    Dan Sindhikara sindhikara@gmail.com
    Does PDB related stuff
    Todo: 
1) Integrate better with AMBER modules (e.g. guessatom, atom radii, etc)



'''
import math

class Atom():
    '''
Atom object
Stores atomic information as would be in a pdb line.

Below is PDB format
1-6 7-11 13-16 17 18-20 22 23-26 27 31-38 39-46 47-54 55-60 61-66 77-78 79-80
Record name Integer Atom Character Residue name Character Integer AChar Real(8.3) Real(8.3) Real(8.3) Real(6.2) Real(6.2) LString(2) element LString(2) charge
"ATOM " serial name altLoc resName chainID resSeq iCode
Atom serial number. Atom name. Alternate location indicator. Residue name. Chain identifier. Residue sequence number. Code for insertion of residues. Orthogonal coordinates for X in Angstroms. Orthogonal coordinates for Y in Angstroms. Orthogonal coordinates for Z in Angstroms. Occupancy. Temperature factor. Element symbol, right-justified. Charge on the atom.
    '''
    def __init__(self,serial=1,name="D",altloc=" ",resname="BCD",chainid=1,resseq=1,icode=" ",coord=[0.0,0.0,0.0],occ=1.0,tfac=0.0,element="D",charge=0.0,hetatm=False,ister=False):
        self.name = name[0:4]
        self.serial=serial
        self.altloc=altloc[0]
        self.resname = resname[0:3]
        self.chainid=str(chainid)[0]
        self.resseq=resseq
        self.icode=icode
        self.coord=coord
        self.occ=occ
        self.tfac=tfac
        self.element=element[0]
        self.charge=charge
        self.hetatm=hetatm
        self.ister=ister # this is the last Atom of the unit. (the next line in the pdb will be say "TER"
    def __str__(self):
        '''
    Returns a PDB format string
        '''
        if not self.hetatm :
            pdbstring="ATOM  %5d %4s%1s%3s %1c%4d%1s   %8.3f%8.3f%8.3f%6.2f%6.2f           %1s%2d\n" % (self.serial,self.name,self.altloc,self.resname,self.chainid,self.resseq,self.icode,self.coord[0],self.coord[1],self.coord[2],self.occ,self.tfac,self.element,self.charge)
        else :
            pdbstring="HETATM%5d %4s%1s%3s %1c%4d%1s   %8.3f%8.3f%8.3f%6.2f%6.2f           %1s%2d\n" % (self.serial,self.name,self.altloc,self.resname,self.chainid,self.resseq,self.icode,self.coord[0],self.coord[1],self.coord[2],self.occ,self.tfac,self.element,self.charge)
        if self.ister :
            pdbstring="%sTER\n" % pdbstring
        return(pdbstring) 

    def makestring(self):
        '''
    Antiquated function same as __str__
        '''
        print("# Warning, antiquated function, __str__ function added")
        return(str(self))

        return(pdbstring)
        
def centermolecule(listofAtoms):
    '''
Given a list of Atom class objects, returns the same list centered
    '''
    center=[0.0,0.0,0.0]
    for atom in listofAtoms:
        for dim in range(3):
            center[dim]+=atom.coord[dim]
    for dim in range(3):
        center[dim]=center[dim]/float(len(listofAtoms))
    for i in range(len(listofAtoms)):
        for dim in range(3):
            listofAtoms[i].coord[dim]-=center[dim]
    return(listofAtoms)


def readcoordinates(pdbfilename):
    ''' Given a pdbfilename, return list of coordinates (2dlist)
This function is less functional than readpdbintoAtoms()
It only stores coordinates
    '''
    pdbfile=open(pdbfilename)
    coords=[]
    for line in pdbfile:
        ident=line[0:6]
        if ident == "ATOM  " or ident == "HETATM" :
            coords.append([float(line[30:38]),float(line[38:46]),float(line[46:54])])
    return(coords)

def readpdbintoAtoms(pdblines,guesselement=False,guesshetatm=False):
    ''' Given a pdb file of arbitrary length parsed into lines,
store the data as a list of "Atom" class objects
    '''

    molecule=[]
    knownelements = ["C","N","O","P","H","S","K","KA"] # put larger ones at the end (so that KA supersedes K)
    knownresidues = ["THR","ILE","ALA","ARG","HIS","HIE","HID","CYS","CYX","LYS","ASP","ASH","GLU","SER","THR","ASN","THR","ASN","GLN","GLY","PRO","LEU","MET","PHE","TRP","TYR","VAL"]
    # The above lists are used to help classify atoms
    if guesselement :
        pass
        #print("Warning! Guesselement only uses single letter atoms")
    terminilist=[]
    for i,line in enumerate(pdblines):
        ident=line[0:6]
        if ident == "ATOM  " or ident == "HETATM" :
            #index in line is one less than column number
            serial=int(line[6:11])
            name=line[12:16]
            altloc=line[16]
            resname=line[17:20]
            chainid=line[21]
            resseq=int(line[22:26])
            icode=line[26]
            coord=[float(line[30:38]),float(line[38:46]),float(line[46:54])]
            occ=float(line[54:60])
            tfac=float(line[60:66])
            if len(line) >= 77 :
                element=line[76:78]
            else:
                element="  "
            if guesselement :
                for knownelement in knownelements:
                    if knownelement in name:
                        element=knownelement
            if len(line) >= 79:
                try :
                    charge=float(line[78:80])
                except :
                    charge=0.0
            else:
                charge=0.0
            hetatm=False
            if ident == "HETATM" :
                hetatm=True
            elif (resname not in knownresidues) and guesshetatm:
                hetatm=True
            molecule.append(Atom(serial,name,altloc,resname,chainid,resseq,icode,coord,occ,tfac,element,charge,hetatm))
        if line[0:3] == "TER" :
            molecule[-1].ister=True
    return(molecule)

def isitinside(coordinate,atoms,buffer=0.0):
    '''
Test if coordinate is inside VDW radius of list of atoms
    '''
    radii={'H' : 0.6, 'O' : 1.6612, 'C' : 1.908, 'S': 2.000, 'N' : 1.8240, 'KA' : 1.7131,  'C0' : 1.7131, 'P' : 2.1000, 'K' : 1.7131}
    inside = False
    for atom in atoms:
        if not atom.ister:
            dist=(coordinate[0]-atom.coord[0])**2+(coordinate[1]-atom.coord[1])**2+(coordinate[2]-atom.coord[2])**2
            dist=math.sqrt(dist)
        
            try:
                 if dist < radii[atom.element[0]]+buffer :
                     inside = True
                     break
            except:
                print("Cannot find radius of atom named: ",atom.element,atom.name)
                exit()
    return(inside)

def isitnear(coordinate,atoms,radius):
    '''
Test if coordinate is within specified radius of listed atoms
    '''
    inside = False
    for atom in atoms:
        dist=(coordinate[0]-atom.coord[0])**2+(coordinate[1]-atom.coord[1])**2+(coordinate[2]-atom.coord[2])**2
        dist=math.sqrt(dist)
        if dist < radius:
            inside = True
            break
    return(inside)


def calcmolvolume(atoms,gridspacing,probesize=0.0):
    import modules.djsgrid as grid
    print("This function is under testing and quite slow. This should be replaced by faster, more reliable code")

    volume=0.0
    print("spacing =",gridspacing)
    gridvolume=gridspacing**3
    print("grid volume is ",gridvolume)
    maxes=[-1000]*3
    mins=[1000]*3
    for atom in atoms:
        if atom.coord[0] < mins[0] :
            mins[0] = atom.coord[0]
        elif atom.coord[0] > maxes[0] :
            maxes[0] = atom.coord[0]
        if atom.coord[1] < mins[1] :
            mins[1] = atom.coord[1]
        elif atom.coord[1] > maxes[1] :
            maxes[1] = atom.coord[1]
        if atom.coord[2] < mins[2] :
            mins[2] = atom.coord[2]
        elif atom.coord[2] > maxes[2] :
            maxes[2] = atom.coord[2]
    # add some buffer
    for i in range(3):
        mins[i]-=5.0
        maxes[i]+=5.0

    spans=[]
    for i in range(3):
        spans.append(int((maxes[i]-mins[i])/gridspacing))

    radmap={'H' : 0.6, 'O' : 1.6612, 'C' : 1.908, 'S': 2.000, 'N' : 1.8240, 'KA' : 1.7131,  'C0' : 1.7131, 'P' : 2.1000, 'K' : 1.7131}
    shelldict={} # construct dictionary of shell indices (these contain all grid points radially from atom center which count as inside the atom
    print(" don't need to compute shellindices more than once, do it for a big one and assign shells")
    allshellindices=grid.precomputeshellindices(int((2.5+probesize)/gridspacing))
    for radkey in radmap: 
        myrad=radmap[radkey]+probesize # effective radius
        maxshellindex=int(myrad/gridspacing+0.5) # largest shell index that counts as inside
        shelldict[radkey]=allshellindices[:maxshellindex]

    occupiedflags=[[[False for i in range(spans[0])] for j in range(spans[1])] for k in range(spans[2])]
    testvol=0
    for shellindex in shelldict["C"]:
        for shellpoint in shellindex:
            testvol+=gridvolume
    print("at this resolution, Carbon volume (plus probe) is",testvol)
    print("by geometry it's", (4.0/3.0)*3.14159 * (radmap["C"]+probesize) **3)

    occupiedpdbfile=open("dumb.pdb","w")
    testedpdbfile=open("tested.pdb","w")
    for atom in atoms:
        myindices=grid.getindicesfromcoord([atom.coord[0],atom.coord[1],atom.coord[2]],mins,[gridspacing]*3)
        #print(atom.coord[0],atom.coord[1],atom.coord[2])
        #print(grid.getcoordfromindices(myindices,mins,[gridspacing]*3))
        myshells=shelldict[atom.element]
        for shell in myshells: # 0th, 1st, 2nd shell, etc
            for shellindex in shell: # points that lay on a shell
                netindex=[]
                for i in range(3): # indices of each point on shell
                    #print("shellindex,myindex[i]=",shellindex[i],myindices[i])
                    netindex.append(shellindex[i]+myindices[i]) # origin of atom + shell index
                    #print(spans)
                #print("netindices=",netindex)
                #print(grid.getcoordfromindices(netindex,mins,[gridspacing]*3))
                coord=grid.getcoordfromindices(netindex,mins,[gridspacing]*3)
                testedpdbfile.write(Atom(coord=coord).makestring())
                if not occupiedflags[netindex[0]][netindex[1]][netindex[2]]:
                    volume+=gridvolume
                    occupiedflags[netindex[0]][netindex[1]][netindex[2]]=True
                    occupiedpdbfile.write(Atom(coord=coord).makestring())
    occupiedpdbfile.close()
    testedpdbfile.close()
    return(volume)






Placevent3/grid.py
# -*- coding: utf-8 -*-

#!/usr/bin/python

'''
Daniel J. Sindhikara
sindhikara@gmail.com
Handles 3D volumetric data
Copyright 2013 Daniel Jon Sindhikara
'''

from __future__ import division
import numpy as np
import os

class Grid:
    '''
Contains volumetric data
    '''

    def __init__(
        self,
        distribution,
        origin,
        gridcount,
        deltas,
        concentration=-1.0,
        ):
        if type(distribution) is list:
            self.distribution = np.array(distribution)
        elif type(distribution) is np.ndarray:
            self.distribution = distribution
        self.origin = origin
        self.gridcount = gridcount
        self.deltas = deltas
        self.concentration = concentration  # in molar

    def getvalue(self, coord): 
        return linearinterpolatevalue(self.distribution, self.origin,
                self.deltas, coord)

    def writedx(self, filename):
        printdxfrom3d(self.distribution, self.origin, self.deltas,
                      self.gridcount, filename)
    def writegrd(self, filename):
        printgrd(self.distribution, self.origin, self.deltas,
                      self.gridcount, filename)

    def nearestgridindices(self, coord):
        '''
        Given a 3D cartesian coordinate, return nearest grid indices
        '''

        gridindices = [int(round((coord[i] - self.origin[i])
                       / self.deltas[i])) for i in range(3)]
        return gridindices

    def coordfromindices(self, indices):
        return [indices[i] * self.deltas[i] + self.origin[i] for i in
                range(3)]

    def coarseRDF(self, coord):
        '''
        Given a 3D cartesian coordinate, return 1D distribution as list, spaced by gridspacing
        '''

        dist = []
        myindices = self.nearestgridindices(coord)
        for shellindex in shellindices:
            avg = 0.0
            try:
                for indexonshell in shellindex:
                    avg += self.distribution[myindices[0]
                            + indexonshell[0]][myindices[0]
                            + indexonshell[1]][myindices[2]
                            + indexonshell[2]]
                avg = avg / len(shellindex)
                dist.append(avg)
            except IndexError:
                break
        return dist

    def interpRDF(
        self,
        coord,
        delta,
        limit,
        concentration = False # If not False, print coordination number
        ):
        '''
        Given 3D cartesian coordinate (list-like object of floats),
        delta (float), and limit (float),
        
        return RDF

        by averaging linear interpolated g(r_vec) on points on spherical shell.
        '''

        spherefilename = \
            os.path.join(os.path.dirname(os.path.realpath(__file__)),
                         'data', 'points', '200.pts')
        spherepoints = [[float(element) for element in line.split()]
                        for line in open(spherefilename).readlines()]
        rdf = []
        volints = [0.]
        for radius in np.arange(0, limit, delta):
            mysum = 0
            for point in spherepoints:
                mysum += self.getvalue([point[dim] * radius
                        + coord[dim] for dim in range(3)])
            rdf.append(float(mysum / len(spherepoints)))
            if radius > 0:
                volints.append(volints[-1] + 4./3. * 3.14159 * (radius**3 - (radius-delta)**3))
        if not concentration:
            return rdf
        else:
            coordnums = [volint * concentration for volint in volints]
            return rdf, coordnums


def rdf23dgrid(
    rdf,
    rdfdelta,
    griddelta,
    gridorigin,
    shellindices,
    ):
    '''
    Convert 1d RDF to 3D grid.
    '''

    gridcount = [len(shellindices) * 2 + 1] * 3  # + 1 is to make it odd, let
    mydist = np.array([[[1.0 for i in range(gridcount[0])] for j in
                      range(gridcount[1])] for k in
                      range(gridcount[2])])

    indexmultiplier = int(griddelta / rdfdelta)  # from 3D grid index to 1D rdf index
    for (i, shell) in enumerate(shellindices):  # loop through shells
        myrdf = rdf[i * indexmultiplier]
        for shellindex in shell:
            distindex = tuple([shellindex[dim] + int(gridcount[dim]
                              / 2) for dim in range(3)])

            # above will put shell 0 at the int(gridcount/2) aka numshells/2 + 1

            mydist[distindex] = myrdf
    return Grid(mydist, gridorigin, gridcount, [griddelta] * 3)


def dx2Grid(dxfilename):
    '''
    Reads a dx file into a Grid class object 
    '''

    # for now Ghetto rigged to use old readdx function

    (distribution, origin, deltas, gridcount) = readdx(dxfilename)
    return Grid(distribution, origin, gridcount, deltas)


def openreadfile(filename):
    '''
Open a file for reading.
Use gzip.open if it's a .gz file.
    '''
    from gzip import open as gopen
    if 'gz' in filename:
        f = gopen(filename,'rb')
    else:
        f = open(filename,'r')
    return(f)    


def readdx(filename):
    '''
    Reads dx files into memory.
    Returns grid data and single 4D array containing dx data [index][xindex][yindex][zindex]
    and the origin, deltas and gridcounts.
    '''
    dxfile = openreadfile(filename)
    dxlines = []
    for i in range(10):  # only need the first few lines to get grid data
        dxlines.append(dxfile.readline())
    dxfile.close()
    gridcounts = []
    origin = []
    deltas = [0, 0, 0]
    startline = 0
    for (i, line) in enumerate(dxlines):
        splitline = line.split()
        if len(splitline) > 2:
            if splitline[1] == '1':
                gridcounts.append(int(splitline[5]))
                gridcounts.append(int(splitline[6]))
                gridcounts.append(int(splitline[7]))

                # print("# gridcounts ",gridcount)

            if splitline[0] == 'origin':
                origin.append(float(splitline[1]))
                origin.append(float(splitline[2]))
                origin.append(float(splitline[3]))
            if splitline[0] == 'delta':
                if float(splitline[1]) > 0:
                    deltas[0] = float(splitline[1])
                if float(splitline[2]) > 0:
                    deltas[1] = float(splitline[2])
                if float(splitline[3]) > 0:
                    deltas[2] = float(splitline[3])
            if splitline[1] == '3':
                numpoints = int(splitline[9])

                # print("# Total number of gridpoints is ",numpoints)

                startline = i + 1
        if startline > 1:
            break
    
    # Read distribution Values
    dxfile = openreadfile(filename)
    dxtext = dxfile.read()
    dxfile.close()
    splittext = dxtext.split()
    del splittext[0:splittext.index('follows') + 1]  # get rid of header text, last word is "follows"
    floats = []
    for element in splittext:
        if len(element) > 0:
            try:
                floats.append(float(element))
            except ValueError:
                pass
    
    # Assign to 3D numpy array
    assert len(floats) == gridcounts[0]*gridcounts[1]*gridcounts[2]
    import numpy as np
    distribution = np.array(floats)
    distribution = np.reshape(distribution, gridcounts)

    return (distribution, origin, deltas, gridcounts)


def readUxDATA(filename, disttypes = ['g']):
    '''
    Reads 'UxDATA' (UVDATA or UUDATA) files into memory.
    Returns dictionary of 3D numpy arrays, and the origin, deltas and gridcounts.
    Dictionary keys refer to <speciesname>.<distributiontype>
    

    Since these files contain many distribution types, the type
    of distributions to keep should be specified. Default 'g' -> g(r)


    First line should look something like: 
    ## <version> <gridcounts> <gridspacing> <number of species>
    ## 20130619        64     0.5     2
   
    '''
    import numpy as np
    distfile = openreadfile(filename)
    filelines = distfile.readlines()
    linesplit = filelines[0].split()
    if len(linesplit) != 5:
        exit("Error! Cannot parse first line of UxDATA file")
    gridcounts = [int(linesplit[2]) for dim in range(3)]
    deltas = [float(linesplit[3]) for dim in range(3)]
    numspecies = int(linesplit[4])

    distcolumns = {}
    origin = False
    for line in filelines[:10]: # Should contain at least one label line
        if 'g(r)' in line:
            labels = line.split()
            for disttype in disttypes:
                for lindex, label in enumerate(labels):
                    if disttype in label[0] and 'k' not in label:
                        distcolumns[disttype] = lindex - 1
        if not origin and '#' not in line:
            origin = [float(element) for element in line.split()[:3]]
                        
    dists = {}
    for linenum, line in enumerate(filelines):
        splitline = line.split()
        if '##<' in line: #New species
            speciesname = splitline[1]
            print("Reading species:", speciesname)
            for key in distcolumns.keys():
                #print key
                distname = speciesname+'.'+key
                dists[distname] = []
                #print distname
        if "#" not in line and len(splitline) > 2:
            # Probably a data line
            for key, colnum in distcolumns.iteritems():
                distname = speciesname+'.'+key
                dists[distname].append(float(splitline[colnum]))
            
    for key, value in dists.iteritems():
        assert len(value) == gridcounts[0]*gridcounts[1]*gridcounts[2]
        myarray = np.array(value)
        dists[key] = np.reshape(myarray, gridcounts, order='F')
    return dists, origin, deltas, gridcounts

def data2Grids(uxdatafilename, disttypes=['g']):
    '''
    Reads a UxDATA into Grid class objects 
    '''
    (distributions, origin, deltas, gridcounts) = readUxDATA(uxdatafilename, disttypes=disttypes)
    grids = {}
    for name, dist in distributions.iteritems():
        grids[name] = Grid(dist, origin, gridcounts, deltas)
    return grids


def readTKRguv(filename):
    '''
    Reads TINKER 'guv' file format
    Returns lists of 3D numpy and the origin, deltas and gridcounts.



    First two lines should look something like:
    16.00000000000000         16.00000000000000         16.00000000000000     
           32           32           32
   
    '''
    print("Reading TINKER style guv file")
    print("Warning, origin set to default 0.0, 0.0, 0.0")
    print("Distributions are unlabeled.")
    print("Assuming distributions start on line 8")
    
    distfile = openreadfile(filename)
    filelines = distfile.readlines()
    
    sidelengths = [float(element) for element in filelines[0].split()]
    gridcounts = [int(element) for element in filelines[1].split()]
    deltas = [sidelengths[dim] / gridcounts[dim] for dim in range(3)]
    origin = [0.0]*3
    numspecies = int(filelines[6])
    print("Found %d species." % (numspecies))
    datalines = [line.split() for line in filelines[7:]]
    assert len(datalines) == gridcounts[0]*gridcounts[1]*gridcounts[2]
    dists = []

    for specnum in range(numspecies):
        dists.append(np.reshape(np.array([float(dataline[specnum]) for dataline in datalines]),
                                gridcounts))
    return dists, origin, deltas, gridcounts


def TKRguv2Grids(filename):
    '''
    Reads a TINKER guv into Grid class objects 
    '''

    (distributions, origin, deltas, gridcounts) = readTKRguv(filename)
    grids = []

    for dist in distributions:
        grids.append(Grid(dist, origin, gridcounts, deltas))
    return grids


def h5ToGrids(h5filename):
    '''
Reads distributions from an MDF hdf5 (h5) file into a dictionary
of Grid class objects.
Returns list of dictionaries:

gridlist[speciesindex]["distributiontype"] = GridObject

    '''
    import tables
    h5file = tables.openFile(h5filename)
    try: 
        offset = h5file.root.parameters_uv.coordinates_offset[:]
    except:
        offset = [0,0,0]
        print("No coordinates offset found. Box will be centered on [0.0, 0.0, 0.0]")
    

    #Get universal parameters
    gridcounts = h5file.root.parameters_uv.num_grid_points[:]
    deltas = [h5file.root.parameters_uv.grid_spacing.read()]*3
    origin = [-deltas[dim]*gridcounts[dim]*0.5+offset[dim] for dim in range(3)]

    speciesnames = [species for molecule in 
                    h5file.root.data_vv.parameters_vv.solvent 
                    for species in molecule.names]
    grids = dict((speciesname, {}) for speciesname in speciesnames)
    for key, value in h5file.root._v_children.iteritems():
        # 
        #print("key (in childen) is ",key)
        if "uv" in key[1:] and len(key) < 5:
            # This leave is an array of distributions           
            # key[1:] is to avoid uvv
            # len(key) < 5 is to avoid /parameters_uv/
            print("Reading:", key)
            value = value[:].T
            for specnum, dist in enumerate(value):
                assert len(dist) == gridcounts[0] * gridcounts[1] * gridcounts[2]
                threeddist = np.reshape(dist,[gridcounts[2], gridcounts[1], gridcounts[0]]).T
#                threeddist = np.reshape(dist,gridcounts)
                grids[speciesnames[specnum]][key] = Grid(threeddist, origin, gridcounts, deltas)
    return grids



def printdxfrom1dzfast(
    values,
    origin,
    delta,
    gridcounts,
    filename,
    ):
    ''' Print a dx file'''
    print("Deprecated Function")
    f = open(filename, 'w')
    f.write("#DX file from Dan's program\n")
    f.write('object 1 class gridpositions counts {0} {1} {2}\n'.format(gridcounts[0],
            gridcounts[1], gridcounts[2]))
    f.write('origin {0} {1} {2}\n'.format(origin[0], origin[1],
            origin[2]))
    f.write('delta {0} 0 0\n'.format(delta[0]))
    f.write('delta 0 {0} 0\n'.format(delta[1]))
    f.write('delta 0 0 {0}\n'.format(delta[2]))
    f.write('object 2 class gridconnections counts {0} {1} {2}\n'.format(gridcounts[0],
            gridcounts[1], gridcounts[2]))
    f.write('object 3 class array type double rank 0 items {0} data follows\n'.format(gridcounts[0]
            * gridcounts[1] * gridcounts[2]))
    for value in values:
        f.write('{0}\n'.format(value))
    f.write('object {0} class field\n'.format(filename))
    f.close()


def printdxfrom3d(
    distribution,
    origin,
    delta,
    gridcounts,
    filename,
    ):
    ''' Print a dx file given a 3d list
    This function is used by Grid objects.
    '''
    f = open(filename, 'w')
    f.write("#DX file from Dan Sindhikara's 'grid' program.\n")
    f.write('object 1 class gridpositions counts {0} {1} {2}\n'.format(gridcounts[0],
            gridcounts[1], gridcounts[2]))
    f.write('origin {0} {1} {2}\n'.format(origin[0], origin[1],
            origin[2]))
    f.write('delta {0} 0 0\n'.format(delta[0]))
    f.write('delta 0 {0} 0\n'.format(delta[1]))
    f.write('delta 0 0 {0}\n'.format(delta[2]))
    f.write('object 2 class gridconnections counts {0} {1} {2}\n'.format(gridcounts[0],
            gridcounts[1], gridcounts[2]))
    f.write('object 3 class array type double rank 0 items {0} data follows\n'.format(gridcounts[0]
            * gridcounts[1] * gridcounts[2]))
    for i in range(gridcounts[0]):
        for j in range(gridcounts[1]):
            for k in range(gridcounts[2]):
                f.write('{0}\n'.format(distribution[i][j][k]))
    f.write('object {0} class field\n'.format(filename))
    f.close()


def printgrd(
    distribution,
    origin,
    deltas,
    gridcounts,
    filename,
    ):
    ''' Print a .grd file readable by Accelrys DS given a 3d list.
    Note: This is not UHBD .grd format!!
    This function is used by Grid objects.
    '''
    boxdims = [(gridcounts[dim]-1)*deltas[dim] for dim in range(3)]
    f = open(filename, 'w')
    f.write("Volumetric data written by grid.py.\n")
    f.write('(1p,e12.5)\n')
    f.write('%8.3f %8.3f %8.3f %8.3f %8.3f %8.3f\n' % (boxdims[0],
                                    boxdims[1], boxdims[2],
                                    90.0, 90.0, 90.0))
    f.write('%5d %5d %5d\n' % (gridcounts[0]-1, gridcounts[1]-1, gridcounts[2]-1))
    
    #pleft and pright are number of grid points left or right of 0,0,0
    pleft = [ origin[dim] / deltas[dim] for dim in range(3)]
    pright = [gridcounts[dim] + pleft[dim] - 2 for dim in range(3)]

    # First number (1 or 3) represents fast variable (x or z)
    f.write('    1 %6d %6d %6d %6d %6d %6d\n' % (pleft[0], pright[0], pleft[1], pright[1], pleft[2], pright[2]))
    for z in range(gridcounts[2]):
        for y in range(gridcounts[1]):
            for x in range(gridcounts[0]):
                f.write('%12.5E\n' % (distribution[x][y][z]))
    f.close()

def getcoordfromindices(indices, origin, deltas):
    '''
Returns coordinates as a length 3 list of floats.
# Optimization on June 10, 2013
    '''
    return [float(indices[i]) * deltas[i] + origin[i] for i in range(3)]


def getindicesfromcoord(coord, origin, deltas):
    indices = []
    for i in range(3):
        indices.append(int((coord[i] - origin[i]) / deltas[i] + 0.5))
    return indices

def precomputeshellindices(maxindex):
    '''return a list of 3d lists containing the indices in successive search shells
i.e. 0 = [0,0,0]
     1 = [[-1,-1,-1],[-1,0,-1],..
    essentially how much to shift i,j,k indices from center to find grid point on shell
at index radius
    This will make evacuation phase faster
    '''

    from math import sqrt
    shellindices = [[[0, 0, 0]]]
    for index in range(1, maxindex):

        indicesinthisshell = []
        for i in range(-index, index + 1):
            for j in range(-index, index + 1):
                for k in range(-index, index + 1):
                    if int(sqrt(i * i + j * j + k * k)) == index:  # I think this will miss some
                        indicesinthisshell.append((i, j, k))
        shellindices.append(tuple(indicesinthisshell))
    return tuple(shellindices)


def createprecomputedindicesjson(numshells=40):
    '''
stores a local file called shells.json
    '''

    shellindices = precomputeshellindices(numshells)
    from json import dump
    import os
    outfile = os.path.join(os.path.dirname(__file__), "data_grid", 'shells.json')
    f = open(outfile, 'w')
    dump(shellindices, f)
    f.close()


def readshellindices():
    import os
    infile = os.path.join(os.path.dirname(__file__),"data_grid", 'shells.json')
    from json import load
    f = open(infile, 'rb')
    shellindices = load(f)
    return shellindices


def getlinearweightsandindices(origin, deltas, coord):
    '''
Given one 3D coordinate, return 8 corner indices and weights
that would allow direct linear interpolation
This subroutine is separated from linearinterpolatevalue to allow
precomputation of weights and indices.

This is using the formula for Trilinear interpolation
Significantly optimized (given typical Python constraints).
    '''

    
    x0i = int((coord[0] - origin[0]) / deltas[0])
    y0i = int((coord[1] - origin[1]) / deltas[1])
    z0i = int((coord[2] - origin[2]) / deltas[2])

    
    x0 = x0i * deltas[0] + origin[0]
    y0 = y0i * deltas[1] + origin[1]
    z0 = z0i * deltas[2] + origin[2]

    x1 = x0 + 1
    y1 = y0 + 1
    z1 = z0 + 1

    #Check if xyz1 is beyond array boundary:
    #x1[np.where(x1==input_array.shape[0])] = x0.max()
    #y1[np.where(y1==input_array.shape[1])] = y0.max()
    #z1[np.where(z1==input_array.shape[2])] = z0.max()

    dx = coord[0] - x0
    dy = coord[1] - y0
    dz = coord[2] - z0
    
    normweights = ((1-dx)*(1-dy)*(1-dz),\
                   dx*(1-dy)*(1-dz),\
                   (1-dx)*dy*(1-dz),\
                   (1-dx)*(1-dy)*dz,\
                   dx*(1-dy)*dz,\
                   (1-dx)*dy*dz,\
                   dx*dy*(1-dz),\
                   dx*dy*dz)

    cindices = ((x0i,y0i,z0i),
             (x0i+1,y0i,z0i),
             (x0i,y0i+1,z0i),
             (x0i,y0i,z0i+1),
             (x0i+1,y0i,z0i+1),
             (x0i,y0i+1,z0i+1),
             (x0i+1,y0i+1,z0i),
             (x0i+1,y0i+1,z0i+1))
    return (normweights, cindices)


def linearinterpolatevalue(
    distribution,
    origin,
    deltas,
    coord,
    ):
    '''given a 3d coordinate, using a linear interpolation from the 8 nearest gridpoints,
estimate the value at that coordinate
    '''

    (weights, cindices) = getlinearweightsandindices(origin, deltas,
            coord)
    value = 0
    for (i, mycindices) in enumerate(cindices):
        try:
            value += distribution[mycindices] * weights[i]
        except:
            print('Failed to find gridpoint at', mycindices)
            print('coordinate=', coord)
            return False
    return value


def calcrdf(
    distribution,
    origin,
    deltas,
    coord,
    deltar=0.1,
    maxradius=20.0,
    numsumgrids=20,
    ):
    '''
Calculates the radial distribution function about a point using the 3d distribution
    '''

    def shellintegral(radius, delta, center):
        sum = 0.0
        count = 0.0
        for i in range(int(2.0 * radius / delta)):
            for j in range(int(2.0 * radius / delta)):
                for k in range(int(2.0 * radius / delta)):
                    x = float(i) * delta - radius
                    y = float(j) * delta - radius
                    z = float(k) * delta - radius
                    thisrad = (x ** 2 + y ** 2 + z ** 2) ** 0.5
                    if thisrad > radius - delta / 2.0 and thisrad \
                        < radius + delta / 2.0:
                        mycoord = [center[0] + x, center[1] + y,
                                   center[2] + z]
                        sum += linearinterpolatevalue(distribution,
                                origin, deltas, mycoord)
                        count += 1.0
        sum = sum / count
        return sum

    radii = [(float(i) + 0.5) * deltar for i in range(int(maxradius
             / deltar))]
    gr = []
    for (i, rad) in enumerate(radii):
        subdelta = deltar * (float(i) + 0.5) / float(numsumgrids)
        gr.append(shellintegral(rad, subdelta, coord))

    return (radii, gr)



